### DISCUSSION TITLE
Should ActionCable be a bulk delivery method?

### DISCUSSION THREAD

**Discussion by ibrahima** [excid3/noticed#502] [Q&A]
My understanding is, a bulk delivery method fires once for all recipients, whereas a regular delivery method fires once for each recipient. Is that correct? If that is the case, then I wonder, should the [ActionCable delivery method](https://github.com/excid3/noticed/blob/main/lib/noticed/delivery_methods/action_cable.rb) be a bulk method? I feel like in this case, the thing you're delivering to is the stream, not the "recipient" directly, so you might get duplicate deliveries if you deliver once to each stream per recipient. But I don't know if I'm not understanding the distinction between regular and bulk delivery methods correctly.

Thank you!

Edit: To clarify, if your stream is set to your recipient, then it behaves as a regular delivery method. But if you're streaming to some other model then it might act as a bulk delivery. Like say, I have a Post record, and I want to notify all commenters on that post. If I set `config.stream = -> { recipient }` and have `recipients -> { [record.author, *record.commenters] }`, then this will automatically figure out who to send the notification to and send it once per recipient. But let's say I have multiple delivery methods in this notifier, and my email notifications are individually addressed to the recipient, but I set `config.stream -> { record }` instead, then it would fire once per recipient user, but then stream to the Post record instead, so it would end up repeating the notification multiple times.

I guess this is kinda moot for ActionCable, because you can just choose to stream to each recipient. But I'm looking at setting up a Turbo Stream custom delivery method, and if I wanna stream to the `Post` object because that's how I've structured my views, I might actually want to use a bulk delivery method. But if my "stream" is for a user, then I would want to treat it as an individual delivery method instead.

Hope this makes sense!

**Comment by ibrahima** [comment]
Hmm, one thing I'm realizing is that despite the fact that there is a separate namespace for bulk delivery methods, you could always pass in the class name explicitly and use it with `bulk_deliver_by`.

https://github.com/excid3/noticed/blob/00b5f96f98e533af29c0983cb5ac096a83a3f57f/app/models/noticed/deliverable/deliver_by.rb#L12

Oh, never mind, that doesn't work because the `perform` methods are different for [regular](https://github.com/excid3/noticed/blob/00b5f96f98e533af29c0983cb5ac096a83a3f57f/lib/noticed/delivery_method.rb#L15) and [bulk](https://github.com/excid3/noticed/blob/00b5f96f98e533af29c0983cb5ac096a83a3f57f/lib/noticed/bulk_delivery_method.rb#L13). I guess in my case I'd need a singular and bulk Turbo Stream custom delivery method.

And yeah, just for my own understanding, the distinction between the two is that bulk is called once per event, regular is called once per notification (i.e. per recipient).

https://github.com/excid3/noticed/blob/00b5f96f98e533af29c0983cb5ac096a83a3f57f/app/jobs/noticed/event_job.rb#L4-L14

**Comment by excid3** [comment]
Think of it this way:

Each user has their own unique stream of notifications. They would subscribe to a channel to listen to only theirs. If you're inserting notifications to the navbar for example, each notification would be broadcast to the unique channel for the user.

Bulk ActionCable notifications could be useful for a situation where everyone's in the same page and watching updates for that page (maybe like a chatroom or something).

Turbo also handles this type of stuff, so it may be worth using Turbo Broadcasts instead of Noticed for these types of things too.

**Reply by ibrahima** [reply]
Ah yeah I am thinking of cases where multiple viewers are looking at the same record and get bulk ActionCable notifications about the resource.

Yeah we do this stuff with Turbo already, I was just trying to understand how I might utilize ActionCable - guess it just depends on what your channel is. I'm actually considering doing the Turbo broadcasts _through_ noticed... not sure if it makes sense yet, but I kinda feel like it might, because I do feel like Turbo Broadcasts are often a type of notification, so gathering those notifications in one place (the Notifier) might make sense.

**Reply by excid3** [reply]
Yeah, also Noticed::Event and Noticed::Notification are both ActiveRecord models, so you can directly add Turbo broadcasts to them. I see the ActionCable delivery method as for other custom things.

**Reply by ibrahima** [reply]
Yeah fair enough. I haven't considered putting the broadcasts on the Notification model itself but that could be interesting as well.

It seems like ultimately ActionCable broadcasts _could_ be a bulk channel, it just depends on what model they're associated with. Right now since the base classes are different I think you'd need to create your own BulkActionCable delivery method or something, which is pretty straightforward if someone needs it. But it might be something that could be included in the gem, but I wouldn't want to confuse anyone so maybe it's better left alone.
